### ISSUE TITLE
General-Purpose bridging mechanism for Grounded Atoms with native types

### ISSUE THREAD

**Issue by luketpeterson** [trueagi-io/hyperon-experimental#351]
Grounded Atoms should have a facility to implement translation code into native types in their host language.

For example, consider a grounded atom that represents an integer.  If a Python client creates a grounded atom by wrapping an integer, it will be a native python integer, and therefore an Integer atom implemented in Rust won't be able to easily unwrap it and treat it as a Rust integer type.

Ultimately we probably want a generalized mechanism for allowing modules to implement plugins for translation in multiple languages.  However, that will be a lot of work.

I think we'll get a good proportion of the benefit by implementing our own internal facility for this translation for a useful set of grounded atoms in the MeTTa std lib.

More discussion on this topic is here:
https://github.com/trueagi-io/hyperon-experimental/issues/333

**Comment by luketpeterson** [comment]
The Python and the Rust stdlibs each contain a number of declarations (numbers, True|False) that conflict and this is already starting to cause some pain.  So I estimate this will need to be tackled in the next couple months as the pain will only increase as the stdlib does.

**Comment by vsbogd** [comment]
Well it is the reason why we didn't want to allow calling Python code from Rust REPL and vice versa before releasing Alpha and do universal REPL after solving an issue with passing the data between Python and Rust.

**Comment by luketpeterson** [comment]
If the current repl is built with Python support, then the Python types are loaded, otherwise the Rust types take precedence.  It's definitely not a great solution, but it is enough to pass the tests for now.

**Comment by vsbogd** [comment]
Following up the offline discussion about bridging Python/Rust/MeTTa types. Lets say we need to write a MeTTa stdlib `plus` function which gets a pair of integers and returns their sum as integer. As this is a grounded function we would like to implement it in Rust and reuse it for the Python grounded numbers. Which types do we need to support in stdlib and how can we call this function on Python integers?

Python represents numbers as big ints. In Rust platform dependent binary integers are used, and big int implemented by a separate library. It is typical to support both big ints and platform dependent integers as two separate types
because both implementations has own strong sides. We could use polymorphic functions to support both types as an arguments of the `plus` function similarly to the example below:
```
(: to_int (-> Atom int))
(= (to_int $x) (case $x ((a a))))

(: to_bigint (-> Atom BigInt))
(= (to_bigint $x) (case $x ((a A) (B B) (P B))))

(: plusi (-> int int int))
(= (plusi $a $b) (int sum))

(: plusb (-> BigInt BigInt BigInt))
(= (plusb $a $b) (bigint sum))

(: plus (-> Atom Atom Atom))
(= (plus $a $b) (plusi (to_int $a) (to_int $b)))
(= (plus $a $b) (plusb (to_bigint $a) (to_bigint $b)))

(: a int)
(: A BigInt)
(: B BigInt)
(: P PythonInt)
(: C SomeType)

!(plus a a)
!(plus B B)
!(plus B P)
!(plus a B)
!(plus a C)
```

Here `plus` is either `plusi` or `plusb` and correct option is choosen by a type of the argument. The second part is a type casting of `PythonInt` type to a `BigInt` which is done using `to_bigint` function. Result of the run is:
```
[(bigint sum), (int sum)] ; a and b are casted to int or biging and both plusi and plusb are called
[(bigint sum)] ; A and B are casted to BigInt and plusb is called
[(bigint sum)] ; A and C are casted to BigInt and plusb is called
[(bigint sum)] ; a and B are casted to BigInt and plusb is called
[(plus a C)] ; not reduced because there is no plus function which can sum SomeType
```

We also can use `Error` or `Empty` to handle type error properly. And inference control to remove duplication of results in a first case (see #448)

**Comment by luketpeterson** [comment]
Thanks for this write up.  I think this is a good solution in general.

However, the superposed result when multiple rules apply: `!(plus a a)` -> `[(bigint sum), (int sum)]` bothers me a little bit.

In another programming language there would just be an order of precedence to the rules so the BigInt conversion and add would never run.  Ideally we can do the same thing and collapse the match on `plus`.

**Comment by vsbogd** [comment]
> However, the superposed result when multiple rules apply: !(plus a a) -> [(bigint sum), (int sum)] bothers me a little bit.

Yes, me to, but I think it is what we can address in an inference control issue #448.

> In another programming language there would just be an order of precedence to the rules so the BigInt conversion and add would never run. Ideally we can do the same thing and collapse the match on plus.

Yes, precedence rules are also a solution. We could use `case` to implement it using precedence, I am not sure it is possible with current interpreter but it should be possible with minimal MeTTa.

**Comment by luketpeterson** [comment]
I don't know much about type systems.  And super-positional types might mean we're in uncharted territory.

However, one observation is that most lossless type conversions are one way.  For example, the `int->BigInt` conversion will never lose information, but BigInts are les flexible than ints (because they can't be converted back to ints)

Ideally it will be possible to build the operational order of precedence from the information about the available type conversion directions.  That will then allow users to add their custom types and converters which will work seamlessly with the types and converters in the stdlib.

**Comment by vsbogd** [comment]
> BigInts are les flexible than ints (because they can't be converted back to ints)

It depends on which stage one is making the conversion. It is not possible statically but dynamically one is able to make conversion taking into account the actual value of the BigInt.

> Ideally it will be possible to build the operational order of precedence from the information about the available type conversion directions. That will then allow users to add their custom types and converters which will work seamlessly with the types and converters in the stdlib.

Yes

**Comment by luketpeterson** [comment]
This work captures a lot of the desiderata.  https://github.com/trueagi-io/hyperon-experimental/pull/597
